File: latticeagi_charter/scalars.py
"""§1.1 scalar grammar: Int, Hash, Public, Signature, Time, Label, ToolName, Resource, ID<P>."""
import base64
import re
import unicodedata
from datetime import datetime, timezone

from .errors import CharterError

RE_LABEL = re.compile(r"^[a-z][a-z0-9_-]{0,63}$")
RE_TOOL = re.compile(r"^[a-z][a-z0-9_]{0,31}\.[a-z][a-z0-9_]{0,31}$")
RE_HEX64 = re.compile(r"^[0-9a-f]{64}$")
RE_ID_CHARS = re.compile(r"^[A-Za-z0-9_-]{21}$")
RE_SEGMENT = re.compile(r"^[A-Za-z0-9_-][A-Za-z0-9_.\-]{0,127}$")
RE_TIME = re.compile(r"^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})\.(\d{3})Z$")
RE_REPO = re.compile(r"^[A-Za-z0-9_.\-]{1,100}/[A-Za-z0-9_.\-]{1,100}$")

ID_PREFIXES = ("cte", "cch", "cgw", "cin", "cpr", "cky", "ccr", "crq", "crl", "cds", "clg")
MAX_SAFE = 9007199254740991


def is_int(v):
    return isinstance(v, int) and not isinstance(v, bool) and 0 <= v <= MAX_SAFE


def is_hash(v):
    return isinstance(v, str) and bool(RE_HEX64.match(v))


is_public = is_hash


def is_label(v):
    return isinstance(v, str) and bool(RE_LABEL.match(v))


def is_tool_name(v):
    return isinstance(v, str) and bool(RE_TOOL.match(v))


def is_signature(v):
    """Canonical unpadded base64url of exactly 64 bytes."""
    if not isinstance(v, str):
        return False
    try:
        raw = base64.urlsafe_b64decode(v + "=" * (-len(v) % 4))
    except Exception:
        return False
    return len(raw) == 64 and base64.urlsafe_b64encode(raw).rstrip(b"=").decode() == v


def is_resource(v):
    if not isinstance(v, str):
        return False
    n = len(v.encode("utf-8"))
    if n < 1 or n > 512:
        return False
    if not re.match(r"^[\x21-\x7e]+$", v):
        return False
    for s in v.split("/"):
        if not RE_SEGMENT.match(s):
            return False
        if s in (".", ".."):
            return False
    return True


def is_id(v, prefix):
    if not isinstance(v, str):
        return False
    head = prefix + "_"
    if not v.startswith(head):
        return False
    return bool(RE_ID_CHARS.match(v[len(head):]))


def _utc_ms(y, mo, d, h, mi, s, ms):
    """Date.UTC semantics: years 0-99 map to 1900+y."""
    if 0 <= y <= 99:
        y += 1900
    try:
        dt = datetime(y, mo, d, h, mi, s, ms * 1000, tzinfo=timezone.utc)
    except ValueError:
        return None
    return int(dt.timestamp() * 1000)


def is_time(v):
    if not isinstance(v, str):
        return False
    m = RE_TIME.match(v)
    if not m:
        return False
    y, mo, d, h, mi, s, ms = (int(x) for x in m.groups())
    if mo < 1 or mo > 12 or d < 1 or d > 31 or h > 23 or mi > 59 or s > 59:
        return False
    got = _utc_ms(y, mo, d, h, mi, s, ms)
    if got is None:
        return False
    # Round-trip: the 0-99 → 1900+y mapping must not have shifted the year,
    # and invalid dates (Feb 30) must not normalize to another day.
    if 0 <= y <= 99:
        return False
    try:
        dt = datetime(y, mo, d, h, mi, s, ms * 1000, tzinfo=timezone.utc)
    except ValueError:
        return False
    return (dt.year, dt.month, dt.day, dt.hour, dt.minute, dt.second) == (y, mo, d, h, mi, s)


def time_ms(t):
    m = RE_TIME.match(t)
    if not m:
        raise CharterError("SCHEMA", "invalid Time")
    y, mo, d, h, mi, s, ms = (int(x) for x in m.groups())
    return _utc_ms(y, mo, d, h, mi, s, ms)


def ms_time(ms):
    if not isinstance(ms, int) or isinstance(ms, bool) or ms < 0 or ms > MAX_SAFE:
        raise CharterError("SCHEMA", "ms out of range")
    dt = datetime.fromtimestamp(ms / 1000, tz=timezone.utc)
    return (f"{dt.year:04d}-{dt.month:02d}-{dt.day:02d}"
            f"T{dt.hour:02d}:{dt.minute:02d}:{dt.second:02d}.{dt.microsecond // 1000:03d}Z")


def is_repository(v):
    return isinstance(v, str) and bool(RE_REPO.match(v))


def is_scalar(v):
    return isinstance(v, str) or is_int(v) or isinstance(v, bool)


def is_nfc(s):
    return s == unicodedata.normalize("NFC", s)


def utf8_bytes(s):
    return len(s.encode("utf-8"))


def require(cond, detail):
    if not cond:
        raise CharterError("SCHEMA", detail)


def make_id(prefix, chars):
    i = f"{prefix}_{chars}"
    if not is_id(i, prefix):
        raise CharterError("SCHEMA", f"invalid {prefix} id")
    return i

File: src/scalars.ts
/** §1.1 scalar grammar: Int, Hash, Public, Signature, Time, Label, ToolName, Resource, ID<P>. */
import { CharterError } from "./errors.ts";

const RE_LABEL = /^[a-z][a-z0-9_-]{0,63}$/;
const RE_TOOL = /^[a-z][a-z0-9_]{0,31}\.[a-z][a-z0-9_]{0,31}$/;
const RE_HASH = /^[0-9a-f]{64}$/;
const RE_PUBLIC = /^[0-9a-f]{64}$/;
const RE_ID_CHARS = /^[A-Za-z0-9_-]{21}$/;
const RE_SEGMENT = /^[A-Za-z0-9_-][A-Za-z0-9_.-]{0,127}$/;
const RE_TIME = /^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})\.(\d{3})Z$/;
const RE_REPO = /^[A-Za-z0-9_.-]{1,100}\/[A-Za-z0-9_.-]{1,100}$/;

export const ID_PREFIXES = ["cte", "cch", "cgw", "cin", "cpr", "cky", "ccr", "crq", "crl", "cds", "clg"] as const;
export type IdPrefix = (typeof ID_PREFIXES)[number];

export const MAX_SAFE = 9007199254740991;

export function isInt(v: unknown): v is number {
  return typeof v === "number" && Number.isSafeInteger(v) && v >= 0 && !Object.is(v, -0);
}

export function isHash(v: unknown): v is string {
  return typeof v === "string" && RE_HASH.test(v);
}

export function isPublic(v: unknown): v is string {
  return typeof v === "string" && RE_PUBLIC.test(v);
}

export function isLabel(v: unknown): v is string {
  return typeof v === "string" && RE_LABEL.test(v);
}

export function isToolName(v: unknown): v is string {
  return typeof v === "string" && RE_TOOL.test(v);
}

/** Canonical unpadded base64url of exactly 64 bytes; decode/re-encode identical. */
export function isSignature(v: unknown): v is string {
  if (typeof v !== "string") return false;
  const raw = Buffer.from(v, "base64url");
  return raw.length === 64 && raw.toString("base64url") === v;
}

export function isResource(v: unknown): v is string {
  if (typeof v !== "string") return false;
  const bytes = Buffer.byteLength(v, "utf8");
  if (bytes < 1 || bytes > 512) return false;
  // ASCII only — excludes percent/backslash/colon/query/unicode by construction.
  if (!/^[\x21-\x7e]+$/.test(v)) return false;
  const segs = v.split("/");
  for (const s of segs) {
    if (!RE_SEGMENT.test(s)) return false;
    if (s === "." || s === "..") return false;
    if (s.includes("..")) {
      // Dot sequences inside a segment: only a leading-dot segment is forbidden by
      // RE_SEGMENT (first char class excludes '.'), and "." / ".." handled above.
    }
  }
  return true;
}

export function isId(v: unknown, prefix: IdPrefix): v is string {
  if (typeof v !== "string") return false;
  const head = prefix + "_";
  if (!v.startsWith(head)) return false;
  return RE_ID_CHARS.test(v.slice(head.length));
}

export function isTime(v: unknown): v is string {
  if (typeof v !== "string") return false;
  const m = RE_TIME.exec(v);
  if (!m) return false;
  const [, ys, mos, ds, hs, mis, ss, mss] = m;
  const y = +ys!, mo = +mos!, d = +ds!, h = +hs!, mi = +mis!, s = +ss!;
  if (mo < 1 || mo > 12 || d < 1 || d > 31 || h > 23 || mi > 59 || s > 59) return false;
  const ms = Date.UTC(y, mo - 1, d, h, mi, s, +mss!);
  if (Number.isNaN(ms)) return false;
  const dt = new Date(ms);
  return (
    dt.getUTCFullYear() === y && dt.getUTCMonth() === mo - 1 &&
    dt.getUTCDate() === d && dt.getUTCHours() === h &&
    dt.getUTCMinutes() === mi && dt.getUTCSeconds() === s
  );
}

export function timeMs(t: string): number {
  const m = RE_TIME.exec(t);
  if (!m) throw new CharterError("SCHEMA", "invalid Time");
  const [, ys, mos, ds, hs, mis, ss, mss] = m;
  return Date.UTC(+ys!, +mos! - 1, +ds!, +hs!, +mis!, +ss!, +mss!);
}

export function msTime(ms: number): string {
  if (!Number.isSafeInteger(ms) || ms < 0) throw new CharterError("SCHEMA", "ms out of range");
  const d = new Date(ms);
  const p = (n: number, w = 2) => String(n).padStart(w, "0");
  return `${p(d.getUTCFullYear(), 4)}-${p(d.getUTCMonth() + 1)}-${p(d.getUTCDate())}` +
    `T${p(d.getUTCHours())}:${p(d.getUTCMinutes())}:${p(d.getUTCSeconds())}.${p(d.getUTCMilliseconds(), 3)}Z`;
}

export function isRepository(v: unknown): v is string {
  return typeof v === "string" && RE_REPO.test(v);
}

export function isScalar(v: unknown): v is string | number | boolean {
  return typeof v === "string" || isInt(v) || typeof v === "boolean";
}

/** NFC check on a decoded string (§1.1: reject, never normalize). */
export function isNfc(s: string): boolean {
  return s === s.normalize("NFC");
}

export function utf8Bytes(s: string): number {
  return Buffer.byteLength(s, "utf8");
}

export function require(cond: unknown, detail: string): asserts cond {
  if (!cond) throw new CharterError("SCHEMA", detail);
}

export function requireInt(v: unknown, what: string): asserts v is number {
  if (!isInt(v)) throw new CharterError("SCHEMA", `${what}: not a safe nonnegative integer`);
}

export function requireHash(v: unknown, what: string): asserts v is string {
  if (!isHash(v)) throw new CharterError("SCHEMA", `${what}: not a 64-hex hash`);
}

export function requireStr(v: unknown, what: string): asserts v is string {
  if (typeof v !== "string" || !isNfc(v)) throw new CharterError("SCHEMA", `${what}: invalid string`);
}

export function makeId(prefix: IdPrefix, chars: string): string {
  const id = `${prefix}_${chars}`;
  if (!isId(id, prefix)) throw new CharterError("SCHEMA", `invalid ${prefix} id`);
  return id;
}
